src/services/resource.js

```diff
@@ -0,0 +1,1297 @@
+    const newResource = await Resource.create({ url });
+    resource = await newResource.get({ plain: true });
+  }
+  return resource;
+};
+
+// Find or create all resource for the list of urls passed.
+const findOrCreateResources = async (urls) => {
+  if (urls === undefined || urls === null || !Array.isArray(urls)) return [];
+  let newURLs;
+  const filteredUrls = [...new Set(urls
+    .filter((url) => typeof url === 'string')
+    .filter((url) => url))];
+  await semaphore.acquire();
+  const currentResources = filteredUrls.length > 0
+    ? await Resource.findAll({
+      where: {
+        url: {
+          [Op.in]: filteredUrls,
+        },
+      },
+      raw: true,
+    }) || []
+    : [];
+  if (currentResources !== undefined
+    || currentResources !== null
+    || currentResources.length !== filteredUrls.length) {
+    const currentResourceURLs = new Set(currentResources
+      .map((currentResource) => currentResource.url));
+    newURLs = filteredUrls.filter((url) => !currentResourceURLs.has(url));
+  }
+  const resources = [
+    ...await Promise.all(newURLs.map(async (url) => {
+      const resource = await Resource.create({ url });
+      return resource.get({ plain: true });
+    })),
+    ...(currentResources || []),
+  ].sort((a, b) => a.id < b.id);
+  semaphore.release();
+  return resources;
+};
+
+// -----------------------------------------------------------------------------
+// Helper functions
+// -----------------------------------------------------------------------------
+// Identify if passed sourceFields contain one or more of the autoDetectedFields.
+const calculateIsAutoDetected = (sourceFields, autoDetectedFields) => (
+  Array.isArray(sourceFields)
+  && sourceFields.length > 0
+  && Array.isArray(autoDetectedFields)
+  && autoDetectedFields.length > 0
+  && autoDetectedFields
+    .filter((field) => sourceFields.includes(field))
+    .length > 0
+);
+
+// Remap the value of an object attribute to a new attribute
+const remapAttributes = (collection, from, to) => (
+  Array.isArray(collection)
+  && collection.length > 0
+  && typeof from === 'string'
+  && typeof to === 'string'
+    ? collection.map((c) => {
+      const result = c;
+      result[to] = result[from];
+      result[from] = undefined;
+      delete result[from];
+      return result;
+    })
+    : []
+);
+
+// Use regex to find all urls within the field
+const collectURLsFromField = (field) => {
+  if (typeof field !== 'string') return [];
+  const urls = field.match(VALID_URL_REGEX) || [];
+  return urls
+    .reduce((matches, match) => {
+      const exists = matches.find((m) => m === match);
+      if (exists) {
+        return matches;
+      }
+      return [...matches, match];
+    }, []);
+};
+
+// Generate a colection of resoruce objects from the list of urls passed
+const resourcesFromField = (
+  genericId,
+  urlsFromField,
+  field,
+  seed = [],
+) => (typeof genericId === 'number'
+  && genericId === parseInt(genericId, 10)
```

Comment:
```js
(typeof genericId === 'number'
  && genericId === parseInt(genericId, 10)
  ```
  
Am I dumb or would a false in the first part mean you never got to the second (ok, sure, desirable) and a true in the first would always be true in the second? Or are you accounting for floats for some reason?

Reply:
I am trying to sanitize the input, how would you suggest I change it.

Reply:
```js
!!(parseInt(genericId, DECIMAL_BASE))
```

Sorry, the only reason it might matter to have less of these is for test coverage

Reply:
marginal, feel free to ignore, was more curious than anything